RPC: Dns rebind protection

holiman opened the pull request:
This pr introduces the flag `--rpcvhosts=<hosts>`, which tells the RPC server about the hostname(s) to expect for incoming requests (defaults to `localhost` unless specified). This protection mechanism also does not bother preventing requests made against a raw ip address, either v4 or v6.  

This prevents attacks like DNS rebinding, which bypasses SOP by simply masquerading as being within the same origin. These attacks do not utilize CORS, since they are not cross-domain.

By explicitly checking the Host-header, the server will not allow requests made against the server with a malicious host domain.

Additionally, this PR fixes a bug introduced in https://github.com/ethereum/go-ethereum/commit/54aeb8e4c0bb9f0e7a6c67258af67df3b266af3d, which caused the `dumpconfig` command to fail with 
```
#build/bin/geth  --maxpeers=0 --nodiscover --rpc dumpconfig
toml: cannot marshal nil log.Logger
``` 

This PR potentially breaks applications which publish RPC ports towards the world; in order to continue doing so, geth needs to be started like this: 
```
geth <flags> --rpcvhosts=infura.io,www.infura.io,...
```

There's also the possibility to use wildcards, effectively disabling the host-protection: 
```
geth <flags> --rpcvhosts=*
```

holiman commented:
Travis errors:
```
ok  	github.com/ethereum/go-ethereum/eth	4.447s	coverage: 25.9% of statements
--- FAIL: TestFastCriticalRestarts (0.00s)
    --- FAIL: TestFastCriticalRestarts/protocol_64_progress_true (4.58s)
    	downloader_test.go:1819: pivot block not locked in after critical section failure
FAIL
``` 
and
```
--- FAIL: TestSimulation (7.30s)
	peer_test.go:426: node 5: failed to exchange bloom filter requirement in round 1. 
		
```
Not related.

GitCop commented:
Thank you for your contribution! Your commits seem to not adhere to the repository coding standards

 * Commit: 0dbce004a954ad7628f0996fd53d1c42b2908780
  * Commit subjects should be kept under 100 characters

Please check the [contribution guidelines](https://github.com/ethereum/go-ethereum#contribution) for more details.

---
This message was auto-generated by https://gitcop.com

karalabe commented on cmd/utils/flags.go:
```diff
@@ -383,6 +383,11 @@ var (
 	}
+	RPCVirtualHostsFlag = cli.StringFlag{
+		Name:  "rpcvhosts",
+		Usage: "Comma separated list of virtual hostnames from which to accept requests (server enforced). Set to * to disable this protection.",
```
Nitpick, do we really need `Set to * to disable this protection.`? Just afraid the help usage text will wrap.

karalabe commented on node/api.go:
```diff
@@ -114,7 +114,7 @@ func (api *PrivateAdminAPI) PeerEvents(ctx context.Context) (*rpc.Subscription,
 
 // StartRPC starts the HTTP RPC API server.
-func (api *PrivateAdminAPI) StartRPC(host *string, port *int, cors *string, apis *string) (bool, error) {
+func (api *PrivateAdminAPI) StartRPC(host *string, port *int, cors *string, apis *string, hosts *string) (bool, error) {
```
Lets call this `vhosts`, otherwise `host` and `hosts` gets ambiguous.

karalabe commented on node/config.go:
```diff
@@ -105,6 +105,15 @@ type Config struct {
+	// By explicitly checking the Host-header, the server will not allow requests
+	// made against the server with a malicious host domain.
+	// Requests using ip address directly are not affected
+	HTTPVirtualHostnames []string `toml:",omitempty"`
```
Nitpick, can we just call this `HTTPVirtualHosts`? I.e. shorted than hostNAMEs and I don't see the added value.

karalabe commented on node/node.go:
```diff
@@ -365,7 +365,7 @@ func (n *Node) stopIPC() {
 
 // startHTTP initializes and starts the HTTP RPC endpoint.
-func (n *Node) startHTTP(endpoint string, apis []rpc.API, modules []string, cors []string) error {
+func (n *Node) startHTTP(endpoint string, apis []rpc.API, modules []string, cors []string, allowedHosts []string) error {
```
`allowedHosts` -> `vhosts` (keep te api consistent).

karalabe commented on node/node.go:
```diff
@@ -393,9 +393,9 @@ func (n *Node) startHTTP(endpoint string, apis []rpc.API, modules []string, cors
 	n.log.Info(fmt.Sprintf("HTTP endpoint opened: http://%s", endpoint))
-
+	n.log.Info(fmt.Sprintf("HTTP config: cors %v, vhosts%v", cors, allowedHosts))
 	// All listeners booted successfully
```
I think these logs were leftovers from the auto log refactor. Could you change them to:

```
n.log.Info("HTTP endpoint opened", "url", fmt.Sprintf("http://%s", endpoint), "cors", strings.Join(cors, ","), "hvosts", strings.Join(vhosts, ","))
```

karalabe commented on rpc/http.go:
```diff
@@ -142,8 +143,11 @@ func (t *httpReadWriteNopCloser) Close() error {
+func NewHTTPServer(cors []string, hosts []string, srv *Server) *http.Server {
+	// Wrap the CORS-handler within a host-handler
+	handler := newCorsHandler(srv, cors)
+	handler = newHostHandler(hosts, handler)
```
Perhaps `hosts` -> `vhosts` and `newHostHandler` -> `newVHostHandler`

karalabe commented on rpc/http.go:
```diff
@@ -195,6 +199,50 @@ func newCorsHandler(srv *Server, allowedOrigins []string) http.Handler {
 		AllowedMethods: []string{http.MethodPost, http.MethodGet},
 		MaxAge:         600,
 		AllowedHeaders: []string{"*"},
+		Debug:          true,
```
What's this for?

karalabe commented on rpc/http.go:
```diff
@@ -195,6 +199,50 @@ func newCorsHandler(srv *Server, allowedOrigins []string) http.Handler {
+
+// ServeHTTP serves JSON-RPC requests over HTTP, implements http.Handler
+func (h *hostHandler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+
```
Pls no empty lines in these cases :P

karalabe commented on rpc/http.go:
```diff
@@ -195,6 +199,50 @@ func newCorsHandler(srv *Server, allowedOrigins []string) http.Handler {
+// since they do in-domain requests against the RPC api. Instead, we can see on the Host-header
+// which domain was used, and validate that against a whitelist.
+type hostHandler struct {
+	AllowedHosts []string
```
Instead of storing a slice, lets store a `map[string]struct{}` instead. That way we can do quick lookups without incurring a linear scanning cost for every single request.

holiman commented:
Addressed the concerns. I also made a last batch ot testing, after the changes were made. 

testscript:
```bash

#!/bin/bash
data='{"jsonrpc":"2.0","method":"web3_clientVersion","params":[],"id":67}'

test(){
	h=$1
	echo "Target $h"
	curl -H "Content-Type: application/json" -H "host: $h" -X POST --data $data 127.0.0.1:8545
}
test localhost
test "127.0.0.1"
test "foobar.com"
test "x.localhost.com"

```

With std settings `build/bin/geth --nodiscover --maxpeers 0 --rpc`:

``` 
#bash testhosts.sh
Target localhost
{"jsonrpc":"2.0","id":67,"result":"Geth/v1.8.0-unstable-543df3f6/linux-amd64/go1.9.2"}
Target 127.0.0.1
{"jsonrpc":"2.0","id":67,"result":"Geth/v1.8.0-unstable-543df3f6/linux-amd64/go1.9.2"}
Target foobar.com
invalid host specified
Target x.localhost.com
invalid host specified
```

With `#build/bin/geth --nodiscover --maxpeers 0 --rpc --rpcvhosts "foobar.com"`

```
#bash testhosts.sh
Target localhost
invalid host specified
Target 127.0.0.1
{"jsonrpc":"2.0","id":67,"result":"Geth/v1.8.0-unstable-543df3f6/linux-amd64/go1.9.2"}
Target foobar.com
{"jsonrpc":"2.0","id":67,"result":"Geth/v1.8.0-unstable-543df3f6/linux-amd64/go1.9.2"}
Target x.localhost.com
invalid host specified

```

With `#build/bin/geth --nodiscover --maxpeers 0 --rpc --rpcvhosts "foobar.com,localhost.com"`

```
#bash testhosts.sh
Target localhost
invalid host specified
Target 127.0.0.1
{"jsonrpc":"2.0","id":67,"result":"Geth/v1.8.0-unstable-543df3f6/linux-amd64/go1.9.2"}
Target foobar.com
{"jsonrpc":"2.0","id":67,"result":"Geth/v1.8.0-unstable-543df3f6/linux-amd64/go1.9.2"}
Target x.localhost.com
invalid host specified
```

With `#build/bin/geth --nodiscover --maxpeers 0 --rpc --rpcvhosts "foobar.com,x.localhost.com"`

```
#bash testhosts.sh
Target localhost
invalid host specified
Target 127.0.0.1
{"jsonrpc":"2.0","id":67,"result":"Geth/v1.8.0-unstable-543df3f6/linux-amd64/go1.9.2"}
Target foobar.com
{"jsonrpc":"2.0","id":67,"result":"Geth/v1.8.0-unstable-543df3f6/linux-amd64/go1.9.2"}
Target x.localhost.com
{"jsonrpc":"2.0","id":67,"result":"Geth/v1.8.0-unstable-543df3f6/linux-amd64/go1.9.2"}

``` 
With wildcard `#build/bin/geth --nodiscover --maxpeers 0 --rpc --rpcvhosts "a,b,*"`:

```
#bash testhosts.sh
Target localhost
{"jsonrpc":"2.0","id":67,"result":"Geth/v1.8.0-unstable-543df3f6/linux-amd64/go1.9.2"}
Target 127.0.0.1
{"jsonrpc":"2.0","id":67,"result":"Geth/v1.8.0-unstable-543df3f6/linux-amd64/go1.9.2"}
Target foobar.com
{"jsonrpc":"2.0","id":67,"result":"Geth/v1.8.0-unstable-543df3f6/linux-amd64/go1.9.2"}
Target x.localhost.com
{"jsonrpc":"2.0","id":67,"result":"Geth/v1.8.0-unstable-543df3f6/linux-amd64/go1.9.2"}

karalabe commented on rpc/http.go:
```diff
@@ -198,3 +201,51 @@ func newCorsHandler(srv *Server, allowedOrigins []string) http.Handler {
+// The virtalHostHandler can prevent DNS rebinding attacks, which do not utilize CORS-headers,
+// since they do in-domain requests against the RPC api. Instead, we can see on the Host-header
+// which domain was used, and validate that against a whitelist.
+type virtalHostHandler struct {
```
`virtalHostHandler` forgot a `u` from virtUal

holiman replied:
Fixed
